File: Python/Tests/get_covariance.py
"""
/*
 * Copyright (C) 2000-2001 QuantLib Group
 * 
 * This file is part of QuantLib.
 * QuantLib is a C++ open source library for financial quantitative
 * analysts and developers --- http://quantlib.sourceforge.net/
 *
 * QuantLib is free software and you are allowed to use, copy, modify, merge,
 * publish, distribute, and/or sell copies of it under the conditions stated 
 * in the QuantLib License.
 *
 * This program is distributed in the hope that it will be useful, but 
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the license for more details.
 *
 * You should have received a copy of the license along with this file;
 * if not, contact ferdinando@ametrano.net
 *
 * QuantLib license is also available at:
 * http://quantlib.sourceforge.net/LICENSE.TXT
*/
"""

""" 
    $Source$
    $Log$
    Revision 1.1  2001/03/15 13:49:35  marmar
    getCovariance function added


"""

import QuantLib
from TestUnit import TestUnit

def initCovariance(vol, corr):
    n = len(vol)
    cov = QuantLib.Matrix(n,n)
    if(n != corr.rows()):
        raise Exception(
            "correlation matrix and volatility vector have different size")
    for i in range(n):
        cov[i][i] = vol[i]*vol[i]
        for j in range(i):
            cov[i][j] = corr[i][j]*vol[i]*vol[j]
            cov[j][i] = cov[i][j]
    return cov


class GetCovarianceTest(TestUnit):
    def doTest(self):

        vol = QuantLib.Array([0.1, 0.5, 1.0])
        corr = QuantLib.Matrix(3, 3)
        corr[0][0] = 1.0; corr[0][1] = 0.2; corr[0][2] = 0.5
        corr[1][0] = 0.2; corr[1][1] = 1.0; corr[1][2] = 0.8
        corr[2][0] = 0.5; corr[2][1] = 0.8; corr[2][2] = 1.0
            
        expectedCov = initCovariance(vol, corr) 
        cov = QuantLib.getCovariance(vol, corr)
            
        numErrors = 0
        for i in range(3):
            for j in range(3):
                if abs(cov[i][j] - expectedCov[i][j]) > 1e-10:
                    numErrors  = numErrors + 1
            
        self.printDetails("QuantLib.getCovariance", cov)            
        self.printDetails("Python.getCovariance  ", expectedCov)            
        return numErrors

if __name__ == '__main__':
    GetCovarianceTest().test("getCovariance")

File: Sources/MonteCarlo/getcovariance.cpp

/*
 * Copyright (C) 2000-2001 QuantLib Group
 *
 * This file is part of QuantLib.
 * QuantLib is a C++ open source library for financial quantitative
 * analysts and developers --- http://quantlib.sourceforge.net/
 *
 * QuantLib is free software and you are allowed to use, copy, modify, merge,
 * publish, distribute, and/or sell copies of it under the conditions stated
 * in the QuantLib License.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the license for more details.
 *
 * You should have received a copy of the license along with this file;
 * if not, contact ferdinando@ametrano.net
 * The license is also available at http://quantlib.sourceforge.net/LICENSE.TXT
 *
 * The members of the QuantLib Group are listed in the Authors.txt file, also
 * available at http://quantlib.sourceforge.net/Authors.txt
*/

/*! \file getcovariance.cpp

    $Source$
    $Log$
    Revision 1.4  2001/04/06 18:46:21  nando
    changed Authors, Contributors, Licence and copyright header

    Revision 1.3  2001/04/04 12:13:24  nando
    Headers policy part 2:
    The Include directory is added to the compiler's include search path.
    Then both your code and user code specifies the sub-directory in
    #include directives, as in
    #include <Solvers1d/newton.hpp>

    Revision 1.2  2001/04/04 11:07:24  nando
    Headers policy part 1:
    Headers should have a .hpp (lowercase) filename extension
    All *.h renamed to *.hpp

    Revision 1.1  2001/03/15 13:48:00  marmar
    getCovariance function added

*/

#include "MonteCarlo/getcovariance.hpp"

namespace QuantLib {

    namespace MonteCarlo {
        using QuantLib::Math::Matrix;

        Matrix getCovariance(const Array &volatilities,
                             const Matrix &correlations){
            int size = volatilities.size();
            QL_REQUIRE(correlations.rows() == size,
                       "getCovariance: volatilities and correlations "
                       "have different size");
            QL_REQUIRE(correlations.columns() == size,
                "getCovariance: correlation matrix is not square");

            Matrix covariance(size,size);
            for(int i = 0; i < size; i++){
                for(int j = 0; j < i; j++){
                    covariance[i][j] = volatilities[i] * volatilities[j] *
                            0.5 * (correlations[i][j] + correlations[j][i]);
                    covariance[j][i] = covariance[i][j];
                }
                covariance[i][i] = volatilities[i] * volatilities[i];
            }
            return covariance;
        }

    }

}
